### PR TITLE
Add option to choose between OMP implementations

### PR DESCRIPTION
**Pull request by larroy**
## Description ##
Adds a build switch to choose OpenMP implementations:

Workaround for #15790 

@karan6181

### CODE REVIEW

**Comment by larroy** [comment]
@mxnet-label-bot add [Build, CMake]

**Comment by larroy** [comment]
@mxnet-label-bot add [Breaking]

File: CMakeLists.txt [hunk 1]
```diff
@@ -433,11 +434,11 @@ if(USE_OPENMP)
...
-  if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/3rdparty/openmp/CMakeLists.txt
-     AND SYSTEM_ARCHITECTURE STREQUAL "x86_64"
-     AND NOT MSVC
-     AND NOT CMAKE_CROSSCOMPILING)
-
+  if(USE_OPENMP_BUNDLED_LLVM AND EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/3rdparty/openmp/CMakeLists.txt
+      AND SYSTEM_ARCHITECTURE STREQUAL "x86_64"
+      AND NOT MSVC
+      AND NOT CMAKE_CROSSCOMPILING)
+    message("Using bundlded LLVM OpenMP")
```
Comment (ChaiBapchya) [inline CMakeLists.txt]: nitpick: bundled

File: CMakeLists.txt [hunk 2]
```diff
@@ -24,6 +24,7 @@ mxnet_option(USE_OLDCMAKECUDA     "Build with old cmake cuda" OFF)
 mxnet_option(USE_NCCL             "Use NVidia NCCL with CUDA" OFF)
 mxnet_option(USE_OPENCV           "Build with OpenCV support" ON)
 mxnet_option(USE_OPENMP           "Build with Openmp support" ON)
+mxnet_option(USE_OPENMP_BUNDLED_LLVM "Build with bundled llvm openmp from 3rdparty" OFF)
```
Comment (apeforest) [inline CMakeLists.txt]: What happens when USE_OPENMP is OFF and USE_OPENMP_BUNDLED_LLVM ON?

Comment (larroy) [reply CMakeLists.txt]: USE_OPENMP takes precedence, so no OMP is used.

Comment (apeforest) [reply CMakeLists.txt]: Is this documented somewhere? Can we set mxnet_option(USE_OPENMP   3RD_PARTY | BUILT_IN)?

I am not an expert in CMake so I can not really decide whether this is the most professional way. LGTM for now. But it's better to have someone from CI team to merge this PR.

**Comment by lebeg** [comment]
> LGTM. But I am not an expert in CMake so I can not really decide whether this is the most proper way. @marcoabreu @lebeg for review.

I think what you were [proposing](https://github.com/apache/incubator-mxnet/pull/15808#discussion_r313211314) would be a better way to handle this. Instead of `ON/OFF` I would define an order in which the Omp versions would be preferred. Like a list `OMP_SYSTEM, OMP_BUNDLED, OMP_OFF`.

For a case like `OMP_SYSTEM` if it's not found it would be an error.

**Comment by larroy** [comment]
@lebeg you might be right, but getting this working is urgent as we can't debug. I would propose doing it incrementally or getting your contribution in this direction.

File: CMakeLists.txt (hunk 2 above)
Comment (larroy) [reply CMakeLists.txt]: I used this for consistency. What you propose is also valid. I would do this incrementally as right now debugging is blocked.

**Comment by apeforest** [comment]
@larroy What's the effort required to implement the multi-selection as you earlier proposed? If this only blocks debugging using CMake and not regular users or developers using Make, I am more leaning towards @lebeg suggestion.

**Comment by larroy** [comment]
CI is blocked now, and it's going to be blocked for a few days. There's problems with Windows and other things. If I make changes I'm afraid this won't be merged for a while. Up to you guys to decide. I think this is urgent enough to have it merged and then refine it. I would say this is a critical fix for being able to debug, as other people need to debug issues related to other tasks. (Make in debug requires updating the linker which is something I would prefer to avoid in my workstation).

The effort to implement your suggestion should be around 30 minutes.

**Comment by marcoabreu** [comment]
I'd also prefer the multi selection.

**Comment by larroy** [comment]
Can you guys review again?

File: CMakeLists.txt [hunk 3]
```diff
@@ -431,10 +433,14 @@ endif()
 
 # ---[ OpenMP
 if(USE_OPENMP)
+  if (NOT OPENMP_FLAVOR)
```
Comment (apeforest) [inline CMakeLists.txt]: Where is OPENMP_FLAVOR from?

File: CMakeLists.txt [hunk 4]
```diff
@@ -16,15 +16,15 @@ if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/build/private/local_config.cmake)
   include(${CMAKE_CURRENT_SOURCE_DIR}/build/private/local_config.cmake)
 endif()
 
+
```
Comment (apeforest) [inline CMakeLists.txt]: As we already agreed, please do a diff before sending the PR. Remove unnecessary blank lines in PR please.

**Comment by apeforest** [comment]
@larroy I thought the suggested change is to use multi-selection for USE_OPENMP not adding additional variable:

It would look like:
```
USE_OPENMP:           "Select with Openmp support"    OMP_SYSTEM |  | OMP_BUNDLED | OMP_OFF)
```

@lebeg Please let me know if my understanding is correct. Thanks

File: CMakeLists.txt (hunk 3 above)
Comment (larroy) [reply CMakeLists.txt]: From cmake command line. Is the selector as requested.

**Comment by larroy** [comment]
@apeforest I think what you propose is not homogeneous with what we have. I think is better to leave an intitial boolean switch and a fine grained selector. It also keeps backwards compatibility.

**Comment by lebeg** [comment]
> @larroy I thought the suggested change is to use multi-selection for USE_OPENMP not adding additional variable:
> 
> It would look like:
> 
> ```
> USE_OPENMP:  "Configure Openmp support"    OMP_SYSTEM |  | OMP_BUNDLED | OMP_OFF)
> ```
> 
> @lebeg Please let me know if my understanding is correct. Thanks

Yes, this is more or less what I had in mind. For the sake of backwards compatibility (and maybe simplicity) keeping the option `ON` would default to `OMP_SYSTEM`.

**Comment by lebeg** [comment]
There seem to be some unintended changes in this PR, otherwise it looks good

**Comment by larroy** [comment]
They are for CI, a change in dmlc is needed first:

https://github.com/dmlc/dmlc-core/pull/558

**Comment by larroy** [comment]
@lebeg please approve then

File: CMakeLists.txt [hunk 5]
```diff
@@ -23,7 +23,7 @@ mxnet_option(USE_CUDA             "Build with CUDA support"   ON)
 mxnet_option(USE_OLDCMAKECUDA     "Build with old cmake cuda" OFF)
 mxnet_option(USE_NCCL             "Use NVidia NCCL with CUDA" OFF)
 mxnet_option(USE_OPENCV           "Build with OpenCV support" ON)
-mxnet_option(USE_OPENMP           "Build with Openmp support" ON)
+mxnet_option(USE_OPENMP           "Build with Openmp support" ON) # OFF | ON | PLATFORM | BUNDLED
```
Comment (apeforest) [inline CMakeLists.txt]: What's the difference between ON and the rest? Can you add some comments here?

File: CMakeLists.txt [hunk 6]
```diff
@@ -452,14 +451,18 @@ if(USE_OPENMP)
...
+    message("Using platform provided OpenMP")
     if(OPENMP_FOUND)
       set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
       set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
       set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
       set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${OpenMP_EXE_LINKER_FLAGS}")
       add_definitions(-DMXNET_USE_OPENMP=1)
     endif()
+  else()
+    message(FATAL_ERROR "USE_OPENMP takes values [PLATFORM, BUNDLED, OFF]")
```
Comment (apeforest) [inline CMakeLists.txt]: "USE_OPENMP takes values [PLATFORM, BUNDLED, OFF, ON]")?

**Comment by apeforest** [comment]
> They are for CI, a change in dmlc is needed first:
> 
> [dmlc/dmlc-core#558](https://github.com/dmlc/dmlc-core/pull/558)

Why is it needed for this PR?

File: CMakeLists.txt [hunk 7]
```diff
@@ -432,14 +432,13 @@ endif()
...
 if(USE_OPENMP)
-  find_package(OpenMP REQUIRED)
   # This should build on Windows, but there's some problem and I don't have a Windows box, so
   # could a Windows user please fix?
-  if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/3rdparty/openmp/CMakeLists.txt
-     AND SYSTEM_ARCHITECTURE STREQUAL "x86_64"
-     AND NOT MSVC
-     AND NOT CMAKE_CROSSCOMPILING)
-
+  if(USE_OPENMP STREQUAL "BUNDLED" AND EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/3rdparty/openmp/CMakeLists.txt
```
Comment (marcoabreu) [inline CMakeLists.txt]: The first check and the others should be separate to throw appropriate error messages

File: CMakeLists.txt [hunk 8]
```diff
@@ -452,14 +451,18 @@ if(USE_OPENMP)
     set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
     add_definitions(-DMXNET_USE_OPENMP=1)
-  else()
+  elseif(USE_OPENMP STREQUAL "PLATFORM" OR USE_OPENMP STREQUAL "ON")
```
Comment (marcoabreu) [inline CMakeLists.txt]: What's the difference between on and platform

**Comment by szha** [comment]
@cjolivier01 if you'd like your change request to be considered a veto, please provide a technical reason. You may refer to the [Apache Voting Process](https://www.apache.org/foundation/voting.html#Veto) for details.

Reopening given that @larroy didn't indicate that he'd like to abandon this change.

File: CMakeLists.txt (hunk 8 above)
Comment (larroy) [reply CMakeLists.txt]: They are the same, for backwards compatibility.

File: CMakeLists.txt (hunk 7 above)
Comment (larroy) [reply CMakeLists.txt]: Can you explain what you mean exactly?

**Comment by larroy** [comment]
Submodule update is necessary as build was failing, there's a fix in the submodule that should be merged soon.

**Comment by cjolivier01** [comment]
this is a repeat of a previously vetoed proposal by Pedro and another by someone he tried to sneak a similar change in, just under a different guise. I really don’t have the energy to fight openly discuss this anymore after two years.

**Comment by cjolivier01** [comment]
in addition, on a fundamental level, flippantly switching between omp libraries while taking nothing else into considerations is extraordinarily hazardous (ie what’s mkl bringing in?).

**Comment by larroy** [comment]
This has repeatedly caused crashes reported by several users which didn't happen when we used the platform OMP which is also the one shipped with binary releases. You are doing a disservice to the community with this attitude, the users are free to choose the OMP implementation that they want.

**Comment by szha** [comment]
@cjolivier01 I didn't realize that. If this is indeed a repetition of a change proposal already vetoed, would you mind linking to it?

I only recall a case where @lebeg attempted to remove the bundled omp, which is different from what @larroy is proposing here. If this is what you're referring to, given the difference, would you mind providing a technical reason for rejecting this change?

**Comment by cjolivier01** [comment]
There is plenty of info out there already on this issue including discussions on dev. I will not be forever to refuge this thing every time Pedro wants to sneak basically the exact same change in every few mo this for two years+ while never actually root-causing the issue (he’s been told what the issue was — not the OMP library’s fault — it’s a big in mxnet — see the issue he is locked out of), but ignores repeatedly.

**Comment by larroy** [comment]
@cjolivier01 stop holding this project hostage. This will be reported to the relevant decision makers in Apache. I hope your powers are revoked due to your track record of destructive actions.

File: CMakeLists.txt (hunk 7 above)
Comment (marcoabreu) [reply CMakeLists.txt]: If someone uses the option "bundled" but the "exists" check fails, they'd receive an error message that they selected an invalid mode.

**Comment by marcoabreu** [comment]
If the submodule update fixes the build, please put it in a separate PR as already stated.

**Comment by marcoabreu** [comment]
While I try to not engage in the discussion above whether this is a match to the previously vetoed PR, I agree that I see some similarities. 

With regards to #15790: This PR doesn't actually solve the problem (which appears to be the same as previously) but gives users a different OMP library choice. That effectively means that the default method we use is still broken and users would have to know that they have to use a different one. I might be no expert in that field, but I agree that to this day we only read something about some assertion being hit and nothing further. From what I can see, there has been no root causing (to the actual root) but a lot based on assumptions. This change seems quite massive (introducing a new OMP Backend choice) while not being based on real data. So i understand the frustration from both sides here: Chris' request for root causing is constantly being denied, while the root cause for pedro seems apparent. But since we are a community and we would like to make a well educated decision as a group, I'd agree with the general request for more root causing in order to create transparency. 

Thinking about it, I agree that this is the same change in a different costume that already had been vetoed with proper technical reasons - but I understand that Chris is getting frustrated by constantly having to repeat his request.

So yeah, I agree with his veto that I'd like to see a deep dive. That should be collected somewhere and if we are at a crossroads, we should bring the topic up for a broader vote on dev@.

**Comment by marcoabreu** [comment]
Another proposal to avoid the difficult personal situation is to hand this issues over to somebody else. Maybe another engineer at Amazon or somebody at Intel or Nvidia would like to be of assistance here. I'm certain that they are also quite capable of debugging concurrency issues. Wdyt?

**Pull request closed without merging**